firedrake/functionspacedata.py

```diff
@@ -0,0 +1,517 @@
+vs VectorElement) can share the PyOP2 Set and Map data.
+"""
+
+from __future__ import absolute_import
+
+import numpy
+import FIAT
+from decorator import decorator
+
+
+from FIAT.finite_element import facet_support_dofs
+from FIAT.tensor_product import horiz_facet_support_dofs, vert_facet_support_dofs
+
+from coffee import base as ast
+
+from pyop2 import op2
+
+from firedrake import dmplex as dm_mod
+from firedrake import halo as halo_mod
+from firedrake import mesh as mesh_mod
+from firedrake.petsc import PETSc
+
+
+__all__ = ("get_shared_data", )
+
+
+@decorator
+def cached(f, mesh, key, *args, **kwargs):
+    """Sui generis caching for a function whose data is
+    associated with a mesh.
+
+    :arg f: The function to cache.
+    :arg mesh: The mesh to cache on (should have a
+        ``_shared_data_cache`` object).
+    :arg key: The key to the cache.
+    :args args: Additional arguments to ``f``.
+    :kwargs kwargs:  Additional keyword arguments to ``f``."""
+    assert hasattr(mesh, "_shared_data_cache")
+    cache = mesh._shared_data_cache[f.__name__]
+    try:
+        return cache[key]
+    except KeyError:
+        result = f(mesh, key, *args, **kwargs)
+        cache[key] = result
+        return result
+
+
+@cached
+def get_global_numbering(mesh, nodes_per_entity):
+    """Get a PETSc Section describing the global numbering.
+
+    This numbering associates function space nodes with topological
+    entities.
+
+    :arg mesh: The mesh to use.
+    :arg nodes_per_entity: a tuple of the number of nodes per
+        topological entity.
+    :returns: A new PETSc Section.
+    """
+    return mesh._plex.createSection([1], nodes_per_entity, perm=mesh._plex_renumbering)
+
+
+@cached
+def get_node_set(mesh, nodes_per_entity):
+    """Get the :class:`node set <pyop2.Set>`.
+
+    :arg mesh: The mesh to use.
+    :arg nodes_per_entity: The number of function space nodes per
+        topological entity.
+    :returns: A :class:`pyop2.Set` for the function space nodes.
+    """
+    global_numbering = get_global_numbering(mesh, nodes_per_entity)
+    # Use a DM to create the halo SFs
+    dm = PETSc.DMShell().create()
+    dm.setPointSF(mesh._plex.getPointSF())
+    dm.setDefaultSection(global_numbering)
+    node_classes = tuple(numpy.dot(nodes_per_entity, mesh._entity_classes))
+    node_set = op2.Set(node_classes, halo=halo_mod.Halo(dm))
+    # Don't need it any more, explicitly destroy.
+    dm.destroy()
+    extruded = bool(mesh.layers)
+    if extruded:
+        node_set = op2.ExtrudedSet(node_set, layers=mesh.layers)
+
+    assert global_numbering.getStorageSize() == node_set.total_size
+    return node_set
+
+
+def get_cell_node_list(mesh, entity_dofs, global_numbering):
+    """Get the cell->node list for specified dof layout.
+
+    :arg mesh: The mesh to use.
+    :arg entity_dofs: The FIAT entity_dofs dict.
+    :arg global_numbering: The PETSc Section describing node layout
```

Comment:
So `global_numbering` is cached on a key (`nodes_per_entity`) that is less distinguishing than `entity_dofs`, therefore it is fine.